bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.auth.oauth2client.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.eclipse.smarthome.auth.oauth2client.internal.Keyword.*;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Date;
+import java.util.UUID;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.util.UrlEncoded;
+import org.eclipse.smarthome.core.auth.client.oauth2.AccessTokenResponse;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthClientService;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthError;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthException;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of OAuthClientService.
+ *
+ * It requires the following services:
+ *
+ * org.eclipse.smarthome.core.storage.Storage (mandatory; for storing grant tokens, access tokens and refresh tokens)
+ *
+ * javax.net.ssl.TrustManager[] (optional; this is used to provide SSL options during HTTPS negotiations with
+ * authorization providers. i.e. during SSLContext.init(...) )
+ *
+ * The OAuthTokens, request parameters are stored and persisted using the "Storage" service.
+ * This allows the token to be automatically refreshed when needed.
+ *
+ * @author Michael Bock - Initial contribution
+ * @author Gary Tse - coauthor
+ *
+ */
+public class OAuthClientServiceImpl implements OAuthClientService {
+
+    public static final String OAUTH_X509_TRUST_MANAGER = "org.eclipse.smarthome.auth.oauth2.X509TrustManager";
+    public static final String STORAGE_SERVICE = "org.eclipse.smarthome.auth.oauth2.StorageService";
+
+    public static final int DEFAULT_TOKEN_EXPIRES_IN_BUFFER_SECOND = 10;
+
+    private transient final Logger logger = LoggerFactory.getLogger(OAuthClientServiceImpl.class);
+
+    private OAuthStoreHandler storeHandler;
+
+    // Constructor params - static
+    private final BundleContext bundleContext;
+    private final String handle;
+    private final int tokenExpiresIn_Buffer;
```

Comment:
I personally feel `tokenExpiresInBuffer` is better that `tokenExpiresIn_Buffer`

Reply:
Same thoughts.  The term "tokenExpiresIn" actually comes from the RFC "tokenExpiresIn".  I wanted to make it more readable.